### DISCUSSION TITLE
Coalescing operator with default nullable struct is treated as default(struct)

### DISCUSSION THREAD

**Discussion by bollhals** [dotnet/csharplang#6016] [Q&A] [answered]
Hi 

I had a strange finding today that caused me to break some test cases of mine, I was unware of this and I was wondering whether this is by intention.

So basically if you do this 
```cs
Console.WriteLine(((int?)null) ?? default)
```
it will write "0", as the default is default(int) and not default(int?). 

Is this a bug or what am I missing here?

Thanks

**Comment by svick** [comment, marked as answer]
This is the expected behavior according to [the spec](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/expressions#1114-the-null-coalescing-operator) (edited for brevity):

> In order of preference, the type of `a ?? b` is `A₀`, `A`, or `B`, where `A` is the type of `a` (provided that `a` has a type), `B` is the type of `b` (provided that `b` has a type), and `A₀` is the underlying type of `A` if `A` is a nullable value type, or `A` otherwise. Specifically, `a ?? b` is processed as follows:
> - […]
> - If `A` is a nullable value type and an implicit conversion exists from `b` to `A₀`, the result type is `A₀`. At run-time, `a` is first evaluated. If `a` is not `null`, `a` is unwrapped to type `A₀`, and this becomes the result. Otherwise, `b` is evaluated and converted to type `A₀`, and this becomes the result.
> - […]

In this case, `A` is `int?` and `A₀` is `int`. And since an implicit conversion exists from `default` to `int`, the result is `(int)default`.

**Reply by bollhals** [reply]
Thanks for the reply!

What is `B` (`default`) in my example? `int` or `int?`

if `B` is `int`, then `B` equals `A₀`
if `B` is `int?`, then `B` does not equal `A₀`, but also can not be implictly converted to `int`. 

So then the question is, why is default in the above case `int` and not `int?`

Or is default without more type info matching the `A₀` type and not the `A` type, thus being 0?

**Reply by svick** [reply]
> What is `B` (`default`) in my example? `int` or `int?`

As I understand it, the `default` literal by itself does not have a type, so `B` does not exist (I have previously edited the parentheticals describing that case out, I guess I shouldn't have). But it can be implicitly converted to any type.

Notice that `B` is not actually used in the bullet point that applies here, so it's okay that it does not exist.

> Or is default without more type info matching the `A₀` type and not the `A` type, thus being 0?

Like the spec says, the order of preference is `A₀`, `A`, `B`. And since `A₀` works, it's the type that's chosen.

**Comment by jnm2** [comment]
I also want to point out that `?? default` is a convenient thing to be able to do with `Nullable<T>` values and have it mean something different from `?? null`. It is idiomatic C# and it compiles into `.GetValueOrDefault()`.
